### PR TITLE
For #18040: no network access startup tests

### PR DESCRIPTION
**Pull request by sv-ohorvath**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by sv-ohorvath** [comment]
@mcomella if you want to have a look at this, maybe there is something you'd like to add

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:129]: nit: it'd be safer to use a boolean or enum

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:134]: Perhaps we should assert that the network is enabled/disabled using the app APIs (`ConnectivityManager` or whatever it is) after running these commands. I'm concerned about a few cases:
- it's also possible to transfer data over ADB, which the firebase data centers may be doing for stability reasons
- we may not be able to execute these commands on certain devices, e.g. firebase may disable shell access for security reasons
- it may take some time for this command to execute so the test isn't actually testing what we want it to test. For example, after enabling wifi, it takes a few seconds before the user actually has network connectivity again.

**Review by mcomella — COMMENTED** [review]
I'm not familiar with the `svc` APIs but, assuming they'll disable the network as we expect, this looks like the approach I was thinking of – thanks! Also, thanks for looking at the STR for https://github.com/mozilla-mobile/fenix/issues/16886 and writing tests specifically for it.

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:134]: Thanks for pointing these out. 

> Perhaps we should assert that the network is enabled/disabled using the app APIs (`ConnectivityManager` or whatever it is) after running these commands. 
> * it may take some time for this command to execute so the test isn't actually testing what we want it to test. For example, after enabling wifi, it takes a few seconds before the user actually has network connectivity again.

Working on this now.

> * we may not be able to execute these commands on certain devices, e.g. firebase may disable shell access for security reasons

What I've found so far is this response from a Firebase engineer, on[ Firebase community](https://firebase-community.slack.com/archives/C1MTSQ5QT/p1611169571083500): 
" williamhester  1 month ago
We don't support adb shell commands, but your Instrumentation test may run shell commands (we can't really stop you)."

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt]: done

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:130]: @mcomella Took your suggestions and made some changes: added Idling resources to wait to be online/offline and fixed some nits, if you care to have another look. Thanks!

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:135]: Why did you decide to use `IdlingResource` as a way to wait for the data to be disabled/enabled? Is it to avoid blocking the test thread and have a mechanism to time out the test if the action never happens?

With the current implementation, I'm concerned that:
- it's very easy to forget to deregister the `IdlingResource` and this state could propagate to other tests
- keeping the `IdlingResource` alive alive longer than it takes to enable/disable the network might have unexpected side effects on the tests (it's _probably_ okay but it seems hard to think through the cases)
- (see my other comment) the `IdlingResource`s are not actually being unregistered

A few alternative implementations I can think of that might help:
- if the IdlingResource deregistered itself once it was idle (though that makes it less re-usable; this might be my least favorite of these implementations)
- implementing the network calls a `Rule` such that the `Rule` registers and deregisters the IdlingResources and give you access to the `setNetworkConnection` APIs
- removing `IdlingResource` and implementing the wait + time out ourselves inside the method to change the network.

The last one seems the simplest to me without unexpected side effects but perhaps it's more error prone? In which case 2 is better but maybe it has unexpected consequences to keep the IdlingResource around? I'm not sure. It seems like a clever and interesting use case for `IdlingResource` but I want to make sure I actually understand why we're using it and the implications as I haven't used it much. :)

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt:1]: nit: license

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt:9]: nit: I like all classes to have kdocs

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt:9]: nit: I like all classes to have kdocs

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt:1]: nit: license

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt:9]: If we decide to continue with `IdlingResource`, this change isn't that important because I don't expect the classes to change much but I notice these two classes have a lot of duplication. Is there a way we can share more of their implementation?

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:34]: The docs don't make it clear if this will actually unregister the resource: it's not clear if we can use a separate instance when registering or unregistering.

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:125]: nit: rename to setNetworkEnabled

**Review by mcomella — COMMENTED** [review]
I want to better understand how we're using `IdlingResource` before we finish the review

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt
Comment (mcomella) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt:9]: To clarify, all classes except the test classes for classes under test (e.g. no need for kdoc on `NoNetworkAccessStartupTest` unless it has a particular need for it).

The reason I want class kdoc is that sometimes the class name doesn't make the class' purpose obvious to someone using the class. For example, `HomeActivity` seems fairly obvious to fenix developer but to a new dev, it begs many questions: if there is a `HomeActivity`, `HomeFragment`, and `BrowserFragment`, where is the `BrowserActivity`? Is this `Activity` the "main" activity of the app? Are there other significant activities or is this app a single-activity type of app?

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt:135]: Yes, we're using Idling Resources in many of our tests to wait just until something is ready, not more or less. If we'd set a strict timeout it could result in flakiness if it takes longer or we'd have tests wait an unnecessary amount of time. I'll take your second suggestion and move both the register & unregister to the test class.

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt]: done

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkConnectedIdlingResource.kt]: done

File: app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt]: done

Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/idlingresource/NetworkDisconnectedIdlingResource.kt]: I removed this class and modified the implementation to work for both

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt]: done

File: app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt]: done

**Comment by sv-ohorvath** [comment]
@mcomella Thanks for the thorough review. I've addressed the change requests, please have another look when you can.

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:42]: nit: I'd think this happens automatically. The `finishActivity` gets called after `base.evaluate`, which I think is where the test content actually runs: https://cs.android.com/androidx/android-test/+/master:runner/rules/java/androidx/test/rule/ActivityTestRule.java;l=549;drc=bf4215f8816fc81db3d105ef2fbc4f748619ab45

So I think this can be removed

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:28]: Not sure if it's what you want here but I added `HomeActivityTestRule(skipOnboarding = true)` if you want to avoid having to do:
```kotlin
homeScreen {
}.dismissOnboarding()
```

It sets the preference before the app starts so that it thinks it's already shown the onboarding screen.

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:54]: I think you need to do `HomeActivityTestRule(launchActivity = false)` to prevent the activity from already being launched by this point

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [inline app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:49]: I'm not sure when we'd want to call `setNetworkEnabled` without waiting for the operation to complete so I feel like this "wait for connected/disconnected" block would better fit inside `setNetworkEnabled` itself. There's also a small duplication of intention (i.e. if `setNetworkEnabled(false)`, you need to correctly call `networkDisconnectedIdlingResource` instead of `*Connected*`) that this would solve.

What do you think?

**Review by mcomella — APPROVED** [review]
Looks good to me! I think there are a few nits we can clean up but they're not too major – I think it'd be fine to land even if we decided not to do them.

Thank you!

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (mcomella) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:54]: > I think it'd be fine to land even if we decided not to do them.

Oh actually, I think this one needs to be addressed for this test to work correctly. However, I trust you can implement the correct solution so no need to get a re-review from me before landing unless you want to.

Comment (mcomella) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:54]: Ah, I added code to throw if there's no network on start up and the test runs as expected (i.e. it fails with that change). I think what might be happening is that we're launching the activity twice. So I guess it'd be an optimization to avoid launching the activity twice (if I understand what's happening correctly) but that this test is working correctly.

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:28]: I just wanted to do some basic navigation after startup.

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:42]: done

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:54]: done

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt]: moved all into setNetworkEnabled.

File: app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt
Comment (sv-ohorvath) [reply app/src/androidTest/java/org/mozilla/fenix/ui/NoNetworkAccessStartupTests.kt:41]: done

**Pull request merged**
